Split loading and managing segments out of SegmentedJournal

deepthidevaki opened the pull request:
## Description

Split loading and managing segments out of `SegmentedJournal`. The goal is not reduce the responsbility of `SegmentedJournal` class.

This PR only moved some code around. In addition, it also removed `synchronized` from a lot of methods. It seems unnecessary, as we already assume a single-writer multiple-readers semantics. Having `synchronized` is not helping us to prevent any race conditions. 

## Related issues

closes #7604 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [x] I've reviewed my own code
* [x] I've written a clear changelist description
* [x] I've narrowly scoped my changes
* [x] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

npepinpe commented on journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java:
```diff
@@ -0,0 +1,184 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class SegmentFileManager {
```
:wrench: From the functionality, it can: `loadExistingSegment` and `createSegment`. I think `SegmentLoader` is fine as a name, since tbh the `JournalSegments` is the one which "manages" segments (e.g. keeps track of them, handles corruption, deletion, etc.).

```suggestion
final class SegmentLoader {
```

npepinpe commented on journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java:
```diff
@@ -0,0 +1,184 @@
+  private final long lastWrittenIndex;
+  private final JournalIndex journalIndex;
+
+  public SegmentFileManager(final long lastWrittenIndex, final JournalIndex journalIndex) {
```
```suggestion
  SegmentFileManager(final long lastWrittenIndex, final JournalIndex journalIndex) {
```

npepinpe commented on journal/src/main/java/io/camunda/zeebe/journal/file/SegmentFileManager.java:
```diff
@@ -0,0 +1,184 @@
+    return buffer.get(0);
+  }
+
+  JournalSegment createSegment(final JournalSegmentDescriptor descriptor, final File segmentFile) {
```
:upside_down_face: I would group the "API" of this class together (i.e. non private functions), since private things are implementation details, so someone looking into this class only has to check the top of the file.

npepinpe commented on journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegments.java:
```diff
@@ -0,0 +1,388 @@
+    this.directory = directory;
+    this.lastWrittenIndex = lastWrittenIndex;
+    this.name = name;
+    segmentLoader = new SegmentFileManager(lastWrittenIndex, this.journalIndex);
```
:wrench: I would rather inject the `SegmentFileManager` instead of constructing it. It might also make it easier to test users of this class (e.g. `SegmentedJournal`) if we can provide a different implementation (though I'd have to look if that's currently the case, I don't have all the tests OTOH now).

npepinpe reviewed (approved):
:rocket: 

I like it a lot! Generally most of these comments/changes don't need a second review, but feel free to request one if you'd like. Please address them before merging though (at least the suggestions/requested changes), even if it's to tell me how wrong I am :smile: 

- :thought_balloon: I wonder why we sometimes use `Segment` (or `Segmented`) as prefix, and sometimes not :smile: It's a total nitpick, not important in the end, but for some reason it sticks out to me.
- :x: Some of the tests in `SegmentedJournalTest` are now clearly for the `JournalSegments` class. Can we split the test class as well? This will reduce the "blast radius" of the tests, since the `JournalSegmentsTest` will be much more focused. For example, tests like `shouldHandlePartiallyWrittenDescriptor`, `shouldHandleCorruptionAtDescriptorWithoutAckedEntries`, and possibly other related ones.
- :x: If `JournalSegments` is meant to be used only within its own package, then we should make its usable API and itself package private, to avoid anybody ever using it outside. Same with other classes in general, e.g. `SegmentFileManager` (though I think that's already the case there).

deepthidevaki commented on journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegments.java:
```diff
@@ -0,0 +1,388 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JournalSegments {
```
It is not actually thread-safe. We expect that only there is only writer. But we don't enforce it. We have rwlocks only to synchronize between readers and the writer in specific operations. `append` for example doesn't use any synchronization because we assume that there is only one writer. We should either document this or enforce it as part of other issue. For now, it doesn't make sense to me that some methods are "synchronized" while some are not. That's why I removed it.

deepthidevaki commented:
> * I wonder why we sometimes use `Segment` (or `Segmented`) as prefix, and sometimes not smile It's a total nitpick, not important in the end, but for some reason it sticks out to me.

The meaning is different `SegmentedJournal` - Segmented is adjective. `SegmentManager` - somthing that manages a segment.

> 
> * x Some of the tests in `SegmentedJournalTest` are now clearly for the `JournalSegments` class. Can we split the test class as well? This will reduce the "blast radius" of the tests, since the `JournalSegmentsTest` will be much more focused. For example, tests like `shouldHandlePartiallyWrittenDescriptor`, `shouldHandleCorruptionAtDescriptorWithoutAckedEntries`, and possibly other related ones.

I split some of the tests into a new class. Some tests remain in the `SegmentedJournalTest` because they are testing "integration", but added a new test in SegmentManagerTest that verifies a lower scope. 

I have addressed other review comments.

npepinpe commented on journal/src/test/java/io/camunda/zeebe/journal/file/SegmentsManagerTest.java:
```diff
@@ -0,0 +1,160 @@
+    final var segments = createSegmentsManager(0);
+
+    // then
+    assertThatNoException().isThrownBy(segments::open);
```
:+1: Uh, I didn't know this one! I only knew the `assertThatCode(() -> {}).doesNotThrowException()`. TIL

deepthidevaki commented:
bors merge
